Linux/O_STM/DOXY/man/man3/OSTM.3
.TH "OSTM" 3 "Sun Apr 1 2018" "C++ Software transactional Memory" \" -*- nroff -*-
.ad l
.nh
.SH NAME
OSTM \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <OSTM\&.h>\fP
.PP
Inherited by \fBBANK\fP, and \fBWAREHOUSE\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "virtual void \fBcopy\fP (std::shared_ptr< \fBOSTM\fP > from, std::shared_ptr< \fBOSTM\fP > to)"
.br
.RI "\fI\fBOSTM\fP required virtual method for deep copy\&. \fP"
.ti -1c
.RI "int \fBGet_Unique_ID\fP () const "
.br
.RI "\fIgetter for unique id \fP"
.ti -1c
.RI "int \fBGet_Version\fP () const "
.br
.RI "\fIgetter for version number \fP"
.ti -1c
.RI "virtual std::shared_ptr< \fBOSTM\fP > \fBgetBaseCopy\fP (std::shared_ptr< \fBOSTM\fP > object)"
.br
.RI "\fI\fBOSTM\fP required virtual method for returning a pointer that is copy of the original pointer\&. \fP"
.ti -1c
.RI "void \fBincrease_VersionNumber\fP ()"
.br
.RI "\fIcommit time increase version number to child object \fP"
.ti -1c
.RI "bool \fBIs_Abort_Transaction\fP () const "
.br
.RI "\fINOT USED YET\&. \fP"
.ti -1c
.RI "bool \fBIs_Can_Commit\fP () const "
.br
.RI "\fINOT USED YET\&. \fP"
.ti -1c
.RI "bool \fBis_Locked\fP ()"
.br
.RI "\fIobject unique lock, try locks mutex return boolean value depends on the lock state \fP"
.ti -1c
.RI "void \fBlock_Mutex\fP ()"
.br
.RI "\fIobject unique lock, locks mutex \fP"
.ti -1c
.RI "\fBOSTM\fP ()"
.br
.RI "\fI\fBOSTM\fP Constructor\&. \fP"
.ti -1c
.RI "\fBOSTM\fP (int _version_number_, int _unique_id_)"
.br
.RI "\fI\fBOSTM\fP Custom Constructor\&. \fP"
.ti -1c
.RI "void \fBSet_Abort_Transaction\fP (bool abortTransaction)"
.br
.RI "\fINOT USED YET\&. \fP"
.ti -1c
.RI "void \fBSet_Can_Commit\fP (bool \fBcanCommit\fP)"
.br
.RI "\fINOT USED YET\&. \fP"
.ti -1c
.RI "void \fBSet_Unique_ID\fP (int \fBuniqueID\fP)"
.br
.RI "\fIsetter for unique id \fP"
.ti -1c
.RI "void \fBSet_Version\fP (int \fBversion\fP)"
.br
.RI "\fIsetter for version number \fP"
.ti -1c
.RI "virtual void \fBtoString\fP ()"
.br
.RI "\fI\fBOSTM\fP required virtual method for display object\&. \fP"
.ti -1c
.RI "void \fBunlock_Mutex\fP ()"
.br
.RI "\fIobject unique lock, unlocks mutex \fP"
.ti -1c
.RI "virtual \fB~OSTM\fP ()"
.br
.RI "\fIDe-constructor\&. \fP"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "int \fBGet_global_Unique_ID_Number\fP ()"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "bool \fBabort_Transaction\fP"
.br
.ti -1c
.RI "bool \fBcanCommit\fP"
.br
.ti -1c
.RI "std::mutex \fBmutex\fP"
.br
.RI "\fIObject built in lock\&. \fP"
.ti -1c
.RI "int \fBuniqueID\fP"
.br
.ti -1c
.RI "int \fBversion\fP"
.br
.ti -1c
.RI "const int \fBZERO\fP = 0"
.br
.RI "\fIMeaningful display for value 0\&. \fP"
.in -1c
.SS "Static Private Attributes"

.in +1c
.ti -1c
.RI "static int \fBglobal_Unique_ID_Number\fP = 0"
.br
.RI "\fIUnique object number increase at object creation\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 17 of file OSTM\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "OSTM::OSTM ()"

.PP
\fBOSTM\fP Constructor\&. Default constructor\&.
.PP
\fBParameters:\fP
.RS 4
\fIversion\fP indicates the version number of the inherited child pointer 
.br
\fIuniqueID\fP is a unique identifier assigned to every object registered in \fBOSTM\fP library 
.br
\fIcanCommit\fP NOT USED YET 
.br
\fIabort_Transaction\fP NOT USED YET 
.RE
.PP

.PP
Definition at line 20 of file OSTM\&.cpp\&.
.PP
References abort_Transaction, canCommit, Get_global_Unique_ID_Number(), uniqueID, version, and ZERO\&.
.PP
.nf
21 {
22     this->version = ZERO;
23     this->uniqueID = Get_global_Unique_ID_Number(); //++global_Unique_ID_Number;
24     this->canCommit = true;
25     this->abort_Transaction = false;
26 }
.fi
.SS "OSTM::OSTM (int _version_number_, int _unique_id_)"

.PP
\fBOSTM\fP Custom Constructor\&. Custom Constructor Used for copy object\&.
.PP
\fBParameters:\fP
.RS 4
\fIversion\fP indicates the version number of the inherited child pointer 
.br
\fIuniqueID\fP is a unique identifier assigned to every object registered in \fBOSTM\fP library 
.br
\fIcanCommit\fP NOT USED YET 
.br
\fIabort_Transaction\fP NOT USED YET 
.RE
.PP

.PP
Definition at line 36 of file OSTM\&.cpp\&.
.PP
References abort_Transaction, canCommit, uniqueID, and version\&.
.PP
.nf
37 {
38    // std::cout << "OSTM COPY CONSTRUCTOR" << global_Unique_ID_Number << std::endl;
39     this->uniqueID = _unique_id_;
40     this->version = _version_number_;
41     this->canCommit = true;
42     this->abort_Transaction = false;
43 }
.fi
.SS "OSTM::~OSTM ()\fC [virtual]\fP"

.PP
De-constructor\&. De-constructor 
.PP
Definition at line 48 of file OSTM\&.cpp\&.
.PP
.nf
48             {
49     //std::cout << "[OSTM DELETE]" << std::endl;
50 }
.fi
.SH "Member Function Documentation"
.PP 
.SS "virtual void OSTM::copy (std::shared_ptr< \fBOSTM\fP > from, std::shared_ptr< \fBOSTM\fP > to)\fC [inline]\fP, \fC [virtual]\fP"

.PP
\fBOSTM\fP required virtual method for deep copy\&. 
.PP
Reimplemented in \fBSLIGO_W\fP, \fBTALLAGH_W\fP, \fBCARLOW_W\fP, \fBCARPHONE_WAREHOUSE\fP, \fBDUNDALK_W\fP, \fBKILKENNY_W\fP, \fBAIB\fP, \fBBOI\fP, \fBBOA\fP, \fBSWBPLC\fP, \fBULSTER\fP, and \fBUNBL\fP\&.
.PP
Definition at line 34 of file OSTM\&.h\&.
.PP
.nf
34 {};  
.fi
.SS "int OSTM::Get_global_Unique_ID_Number ()\fC [private]\fP"
Returning global_Unique_ID_Number to the constructor
.PP
If global_Unique_ID_Number equals to 10000000 then reset back to ZERO, to make sure the value of global_Unique_ID_Number never exceed the MAX_INT value 
.PP
Definition at line 56 of file OSTM\&.cpp\&.
.PP
References global_Unique_ID_Number\&.
.PP
Referenced by OSTM()\&.
.PP
.nf
56                                       {
57     if(global_Unique_ID_Number > 10000000)
58         global_Unique_ID_Number = 0;
59     return ++global_Unique_ID_Number;
60 }
.fi
.SS "int OSTM::Get_Unique_ID () const"

.PP
getter for unique id 
.PP
\fBParameters:\fP
.RS 4
\fIuniqueID\fP int 
.RE
.PP

.PP
Definition at line 73 of file OSTM\&.cpp\&.
.PP
References uniqueID\&.
.PP
Referenced by toString(), ULSTER::toString(), UNBL::toString(), SWBPLC::toString(), BOA::toString(), BOI::toString(), AIB::toString(), KILKENNY_W::toString(), CARLOW_W::toString(), DUNDALK_W::toString(), SLIGO_W::toString(), TALLAGH_W::toString(), and CARPHONE_WAREHOUSE::toString()\&.
.PP
.nf
74 {
75     return uniqueID;
76 }
.fi
.SS "int OSTM::Get_Version () const"

.PP
getter for version number 
.PP
\fBParameters:\fP
.RS 4
\fIversion\fP int 
.RE
.PP

.PP
Definition at line 89 of file OSTM\&.cpp\&.
.PP
References version\&.
.PP
Referenced by toString(), ULSTER::toString(), UNBL::toString(), SWBPLC::toString(), BOA::toString(), BOI::toString(), AIB::toString(), KILKENNY_W::toString(), CARLOW_W::toString(), DUNDALK_W::toString(), SLIGO_W::toString(), TALLAGH_W::toString(), and CARPHONE_WAREHOUSE::toString()\&.
.PP
.nf
90 {
91     return version;
92 }
.fi
.SS "virtual std::shared_ptr<\fBOSTM\fP> OSTM::getBaseCopy (std::shared_ptr< \fBOSTM\fP > object)\fC [inline]\fP, \fC [virtual]\fP"

.PP
\fBOSTM\fP required virtual method for returning a pointer that is copy of the original pointer\&. 
.PP
Reimplemented in \fBSLIGO_W\fP, \fBTALLAGH_W\fP, \fBCARLOW_W\fP, \fBCARPHONE_WAREHOUSE\fP, \fBDUNDALK_W\fP, \fBKILKENNY_W\fP, \fBAIB\fP, \fBBOA\fP, \fBBOI\fP, \fBSWBPLC\fP, \fBULSTER\fP, and \fBUNBL\fP\&.
.PP
Definition at line 38 of file OSTM\&.h\&.
.PP
.nf
38 {};//std::cout << "[OSTM GETBASECOPY]" << std::endl;};
.fi
.SS "void OSTM::increase_VersionNumber ()"

.PP
commit time increase version number to child object 
.PP
\fBParameters:\fP
.RS 4
\fIversion\fP int 
.RE
.PP

.PP
Definition at line 97 of file OSTM\&.cpp\&.
.PP
References version\&.
.PP
Referenced by toString()\&.
.PP
.nf
98 {
99     this->version += 1;
100 }
.fi
.SS "bool OSTM::Is_Abort_Transaction () const"

.PP
NOT USED YET\&. 
.PP
\fBParameters:\fP
.RS 4
\fIabort_Transaction\fP boolean 
.RE
.PP

.PP
Definition at line 126 of file OSTM\&.cpp\&.
.PP
References abort_Transaction\&.
.PP
Referenced by toString()\&.
.PP
.nf
126                                       {
127     return abort_Transaction;
128 }
.fi
.SS "bool OSTM::Is_Can_Commit () const"

.PP
NOT USED YET\&. 
.PP
\fBParameters:\fP
.RS 4
\fIcanCommit\fP boolean 
.RE
.PP

.PP
Definition at line 112 of file OSTM\&.cpp\&.
.PP
References canCommit\&.
.PP
Referenced by toString()\&.
.PP
.nf
112                                {
113     return canCommit;
114 }
.fi
.SS "bool OSTM::is_Locked ()"

.PP
object unique lock, try locks mutex return boolean value depends on the lock state 
.PP
\fBParameters:\fP
.RS 4
\fImutex\fP std::mutex 
.RE
.PP

.PP
Definition at line 147 of file OSTM\&.cpp\&.
.PP
References mutex\&.
.PP
Referenced by toString()\&.
.PP
.nf
147                     {
148     return this->mutex\&.try_lock();
149 }
.fi
.SS "void OSTM::lock_Mutex ()"

.PP
object unique lock, locks mutex 
.PP
\fBParameters:\fP
.RS 4
\fImutex\fP std::mutex 
.RE
.PP

.PP
Definition at line 133 of file OSTM\&.cpp\&.
.PP
References mutex\&.
.PP
Referenced by toString()\&.
.PP
.nf
133                       {
134     this->mutex\&.lock();
135 }
.fi
.SS "void OSTM::Set_Abort_Transaction (bool abortTransaction)"

.PP
NOT USED YET\&. 
.PP
\fBParameters:\fP
.RS 4
\fIabort_Transaction\fP boolean 
.RE
.PP

.PP
Definition at line 119 of file OSTM\&.cpp\&.
.PP
References abort_Transaction\&.
.PP
Referenced by toString()\&.
.PP
.nf
119                                                       {
120     this->abort_Transaction = abortTransaction;
121 }
.fi
.SS "void OSTM::Set_Can_Commit (bool canCommit)"

.PP
NOT USED YET\&. 
.PP
\fBParameters:\fP
.RS 4
\fIcanCommit\fP boolean 
.RE
.PP

.PP
Definition at line 105 of file OSTM\&.cpp\&.
.PP
References canCommit\&.
.PP
Referenced by toString()\&.
.PP
.nf
105                                         {
106     this->canCommit = canCommit;
107 }
.fi
.SS "void OSTM::Set_Unique_ID (int uniqueID)"

.PP
setter for unique id 
.PP
\fBParameters:\fP
.RS 4
\fIuniqueID\fP int 
.RE
.PP

.PP
Definition at line 66 of file OSTM\&.cpp\&.
.PP
References uniqueID\&.
.PP
Referenced by ULSTER::copy(), UNBL::copy(), SWBPLC::copy(), BOA::copy(), AIB::copy(), BOI::copy(), and toString()\&.
.PP
.nf
66                                      {
67     this->uniqueID = uniqueID;
68 }
.fi
.SS "void OSTM::Set_Version (int version)"

.PP
setter for version number 
.PP
\fBParameters:\fP
.RS 4
\fIversion\fP int 
.RE
.PP

.PP
Definition at line 81 of file OSTM\&.cpp\&.
.PP
References version\&.
.PP
Referenced by toString()\&.
.PP
.nf
82 {
83     this->version = version;
84 }
.fi
.SS "virtual void OSTM::toString ()\fC [inline]\fP, \fC [virtual]\fP"

.PP
\fBOSTM\fP required virtual method for display object\&. 
.PP
Reimplemented in \fBCARPHONE_WAREHOUSE\fP, \fBSLIGO_W\fP, \fBTALLAGH_W\fP, \fBCARLOW_W\fP, \fBDUNDALK_W\fP, \fBKILKENNY_W\fP, \fBAIB\fP, \fBBOA\fP, \fBBOI\fP, \fBSWBPLC\fP, \fBULSTER\fP, and \fBUNBL\fP\&.
.PP
Definition at line 42 of file OSTM\&.h\&.
.PP
References canCommit, Get_Unique_ID(), Get_Version(), increase_VersionNumber(), Is_Abort_Transaction(), Is_Can_Commit(), is_Locked(), lock_Mutex(), Set_Abort_Transaction(), Set_Can_Commit(), Set_Unique_ID(), Set_Version(), uniqueID, unlock_Mutex(), and version\&.
.PP
.nf
42 {};
.fi
.SS "void OSTM::unlock_Mutex ()"

.PP
object unique lock, unlocks mutex 
.PP
\fBParameters:\fP
.RS 4
\fImutex\fP std::mutex 
.RE
.PP

.PP
Definition at line 140 of file OSTM\&.cpp\&.
.PP
References mutex\&.
.PP
Referenced by toString()\&.
.PP
.nf
140                         {
141     this->mutex\&.unlock();
142 }
.fi
.SH "Member Data Documentation"
.PP 
.SS "bool OSTM::abort_Transaction\fC [private]\fP"

.PP
Definition at line 108 of file OSTM\&.h\&.
.PP
Referenced by Is_Abort_Transaction(), OSTM(), and Set_Abort_Transaction()\&.
.SS "bool OSTM::canCommit\fC [private]\fP"

.PP
Definition at line 104 of file OSTM\&.h\&.
.PP
Referenced by Is_Can_Commit(), OSTM(), Set_Can_Commit(), and toString()\&.
.SS "int OSTM::global_Unique_ID_Number = 0\fC [static]\fP, \fC [private]\fP"

.PP
Unique object number increase at object creation\&. 
.PP
Definition at line 112 of file OSTM\&.h\&.
.PP
Referenced by Get_global_Unique_ID_Number()\&.
.SS "std::mutex OSTM::mutex\fC [private]\fP"

.PP
Object built in lock\&. 
.PP
Definition at line 120 of file OSTM\&.h\&.
.PP
Referenced by is_Locked(), lock_Mutex(), and unlock_Mutex()\&.
.SS "int OSTM::uniqueID\fC [private]\fP"

.PP
Definition at line 100 of file OSTM\&.h\&.
.PP
Referenced by Get_Unique_ID(), OSTM(), Set_Unique_ID(), and toString()\&.
.SS "int OSTM::version\fC [private]\fP"

.PP
Definition at line 96 of file OSTM\&.h\&.
.PP
Referenced by Get_Version(), increase_VersionNumber(), OSTM(), Set_Version(), and toString()\&.
.SS "const int OSTM::ZERO = 0\fC [private]\fP"

.PP
Meaningful display for value 0\&. 
.PP
Definition at line 116 of file OSTM\&.h\&.
.PP
Referenced by OSTM()\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for C++ Software transactional Memory from the source code\&.


Linux/OSTM/OSTM.cpp
/* 
 * File:   OSTM.cpp
 * Author: Zoltan Fuzesi C00197361, 
 * IT Carlow, Software Engineering, 
 *
 * Supervisor : Joe Kehoe, 
 *
 * C++ Software Transactional Memory, 
 * 
 * Created on December 18, 2017, 2:09 PM
 * OSTM base class function declarations.
 */

#include "OSTM.h"

int OSTM::global_Unique_ID_Number = 0;

/*!
 * \brief @21 Default constructor
 */
OSTM::OSTM()
{
    /** @24 Integer field <version> indicates the version number of the inherited child object */
    this->version = ZERO;
    /** @26 Integer field <uniqueID> is a unique identifier assigned to every object registered in OSTM library */
    this->uniqueID = Get_global_Unique_ID_Number();
    /** @28 Boolean value <canCommit> to determine the object can or cannot commit */
    this->canCommit = true;
    /** @30 Boolean field <abort_Transaction> to determine the object can or cannot commit */
    this->abort_Transaction = false;
}


/*!
 * \brief @39 Custom Constructor Used to copying objects
 * \param _version_number_  Integer value used to create a copy of the object with the actual version
 * \param _unique_id_ Integer value used to create a copy of the object with the original unique ID
 */
OSTM::OSTM(int _version_number_, int _unique_id_)
{
    /** @42 Integer field <version> indicates the version number of the inherited child object */
    this->uniqueID = _unique_id_;
    /** @44 Integer field <uniqueID> is a unique identifier assigned to every object registered in OSTM library */
    this->version = _version_number_;
    /** @46 Boolean value <canCommit> to determine the object can or cannot commit */
    this->canCommit = true;
    /** @48 Boolean value <abort_Transaction> to determine the object can or cannot commit */
    this->abort_Transaction = false;
}

/*!
 * @54 Default De-constructor
 */
OSTM::~OSTM() {
    /** Delete the object. */
}
/*!
 * \brief @61 Get_global_Unique_ID_Number function, If <global_Unique_ID_Number> equals to 10000000 then reset back to ZERO, to make sure the value of 
 * global_Unique_ID_Number never exceed the MAX_INT value
 */
int OSTM::Get_global_Unique_ID_Number() {
    /** @64 Checking the global_Unique_ID_Number */
    if(global_Unique_ID_Number > 10000000)
        /** @65 Reset global_Unique_ID_Number to ZERO*/
        global_Unique_ID_Number = 0;
    /** @67 return static global_Unique_ID_Number */
    return ++global_Unique_ID_Number;
}

/*!
 * \brief @75 Function <Set_Unique_ID> setter for uniqueID private field
 * \param uniqueID int Every object inherit from OSTM class will include a version number that is unique for every object.
 * The STM library used this value to find object within the transaction to make changes or comparism ith them.
 */
void OSTM::Set_Unique_ID(int uniqueID) {
    /** @77 set object uniqueID to parameter integer value */
    this->uniqueID = uniqueID;
}
/*!
 * \brief @82 Function <Get_Unique_ID> getter for uniqueID private field
 */
int OSTM::Get_Unique_ID() const
{
    /** @85 return Object uniqueID */
    return uniqueID;
}
/*!
 * \brief @92 Function <Set_Version> setter for version private filed
 * \param version integer The verion number ZERO by default when the object created. When a transaction make changes with the object, then the version number
 * will be increased, to indicate the changes on the object.
 */
void OSTM::Set_Version(int version)
{
    /** @95 set object version to parameter integer value */
    this->version = version;
}
/*!
 * \brief @100 Function <Get_Version> setter for version private filed
 */
int OSTM::Get_Version() const
{
    /** return object version number */
    return version;
}
/*!
 * \brief @108 Function <increase_VersionNumber> commit time increase the version number associated with the object
 */
void OSTM::increase_VersionNumber()
{
    /** @111 increase object version number */
    this->version += 1;
}
/*!
 * \brief @117 Function <Set_Can_Commit> setter for canCommit private filed
 * \param canCommit Boolean value to determine the object can or cannot commit
 */
void OSTM::Set_Can_Commit(bool canCommit) {
    /** @119 set canCommit object variable to parameter boolean value*/
    this->canCommit = canCommit;
}
/*!
 * \brief @124 Function <Is_Can_Commit> boolean function to determin the object can comit or need to roolback.
 */
bool OSTM::Is_Can_Commit() const {
    /** @126 return canCommit boolean value TRUE/FALSE */
    return canCommit;
}
/*!
 * \brief @132 Function <Set_Abort_Transaction> setter for abortTransaction private filed
 * \param abortTransaction Boolean to determine the object can or cannot commit
 */
void OSTM::Set_Abort_Transaction(bool abortTransaction) {
    /** @134 set abort_Transaction object variable to parameter boolean value  */
    this->abort_Transaction = abortTransaction;
}
/*!
 * \brief @140 Function <Is_Abort_Transaction> return boolean value stored in the <abortTransaction> private filed
 * \param abort_Transaction Boolean to determine the object can or cannot commit
 */
bool OSTM::Is_Abort_Transaction() const {
    /** @142 return abort_Transaction object boolean value */
    return abort_Transaction;
}
/*!
 * \brief @145 Function <lock_Mutex> setter for mutex to lock the object
 */
void OSTM::lock_Mutex() {
    /** @149 Locking the mutex*/
    this->mutex.lock();
}
/*!
 * \brief @154 Function <unlock_Mutex> setter for mutex to unlock the object
 */
void OSTM::unlock_Mutex() {
    /** @156 Locking the mutex */
    this->mutex.unlock();
}
/*!
 * \brief @162 Function <is_Locked> Boolean function to try lock the object. If the object
 * not locked then locks and return True it otherwise return False. 
 */
bool OSTM::is_Locked(){
    /** @164 Try to unlock the mutex, return TRUE if the lock was acquired successfully, otherwise return FALSE */
    return this->mutex.try_lock();
}